components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesProvider.cs

```diff
@@ -0,0 +1,159 @@
+            _lock.Wait();
+            try {
+                using (var fileStream = new FileStream(
+                    _legacyCliModel.PublishedNodesFile,
+                    FileMode.Open,
+                    FileAccess.Read,
+                    FileShare.Read
+                 )) {
+                    return fileStream.ReadAsString(Encoding.UTF8);
+                }
+            }
+            catch (Exception e) {
+                _logger.Error(e, "Failed to read content of published nodes file from \"{path}\"",
+                    _legacyCliModel.PublishedNodesFile);
+                throw;
+            }
+            finally {
+                _lock.Release();
+            }
+
+        }
+
+        /// <summary>
+        /// Write new content to published nodes file
+        /// </summary>
+        /// <param name="content"> Content to be written. </param>
+        /// <param name="disableRaisingEvents"> If set FileSystemWatcher notifications will be disabled while updating the file.</param>
+        public void WriteContent(string content, bool disableRaisingEvents = false) {
+            _lock.Wait();
+            try {
+                // Store current state.
+                var eventState = FileSystemWatcher.EnableRaisingEvents;
+
+                if (disableRaisingEvents) {
+                    FileSystemWatcher.EnableRaisingEvents = false;
+                }
+
+                try {
+                    using (var fileStream = new FileStream(
+                        _legacyCliModel.PublishedNodesFile,
+                        FileMode.Open,
+                        FileAccess.Write,
+                        // We will require that there is no other process using the file.
+                        FileShare.None
+                     )) {
+                        fileStream.Write(Encoding.UTF8.GetBytes(content));
+                    }
+                }
+                catch (IOException e) {
+
+                    // We will fall back to writing with ReadWrite access.
+                    try {
+                        using (var fileStream = new FileStream(
+                            _legacyCliModel.PublishedNodesFile,
+                            FileMode.Open,
+                            FileAccess.Write,
+                            // Relaxing requirements.
+                            FileShare.ReadWrite
+                         )) {
+                            fileStream.Write(Encoding.UTF8.GetBytes(content));
+                        }
+                    }
+                    catch (Exception) {
+                        // Report and raise original exception if fallback also failed.
+                        _logger.Error(e, "Failed to update published nodes file at \"{path}\"",
+                            _legacyCliModel.PublishedNodesFile);
+
+                        throw e;
+                    }
+                }
+
+                // Retore state.
+                if (disableRaisingEvents) {
```

Comment:
>disableRaisingEvents [](http://example.com/codeflow?start=20&length=20)

perhaps we should restore the raise events setting in the finally block

Reply:
Good point.

Reply:
Moved it to `finally` block.